File: Shared/Blog.Core/Helpers/ExceptionHelper.cs
// <copyright file="ExceptionHelper.cs" company="PlaceholderCompany">
// Copyright (c) PlaceholderCompany. All rights reserved.
// </copyright>

namespace Blog.Core.Helpers;

using System;
using System.Text;

/// <summary>
/// Exception helper.
/// </summary>
public static class ExceptionHelper
{
    /// <summary>
    /// Gets the messages.
    /// </summary>
    /// <param name="exception">The exception.</param>
    /// <returns>string.</returns>
    public static string GetMessages(Exception exception) => exception == null ? "Exception is NULL" : ExceptionHelper.GetErrorMessage(exception);

    /// <summary>
    /// Gets the error message.
    /// </summary>
    /// <param name="exception">The exception.</param>
    /// <returns>string.</returns>
    private static string GetErrorMessage(Exception exception)
    {
        var stringBuilder = new StringBuilder();
        stringBuilder.AppendLine(exception.Message);
        if (exception.InnerException != null)
        {
            stringBuilder.AppendLine(ExceptionHelper.GetErrorMessage(exception.InnerException));
        }

        return stringBuilder.ToString();
    }
}

File: Shared/Blog.FSharp.Core/Helpers/ExceptionHelper.fs
namespace Blog.FSharp.Core.Helpers

open System
open System.Text

/// <summary>
/// Exception helper.
/// </summary>
module ExceptionHelper =

    /// <summary>
    /// Recursively gets the error message from the exception and its inner exceptions.
    /// </summary>
    /// <param name="ex">The exception.</param>
    /// <returns>All messages as a single string.</returns>
    let rec private getErrorMessage (ex: Exception) : string =
        let sb = StringBuilder()
        sb.AppendLine(ex.Message) |> ignore
        if not (isNull ex.InnerException) then
            // Явно вказуємо, що результат getErrorMessage – string
            let innerMsg: string = getErrorMessage ex.InnerException
            sb.AppendLine(innerMsg) |> ignore
        sb.ToString()

    /// <summary>
    /// Gets the messages.
    /// </summary>
    /// <param name="ex">The exception.</param>
    /// <returns>string.</returns>
    let getMessages (ex: Exception) : string =
        if isNull ex then "Exception is NULL"
        else getErrorMessage ex